File: iOS/ssabree/ssabree/Features/MyGroup/MemberManageViewModel.swift
import Foundation
import Observation

struct MemberManageUiState {
    var members: [GroupMemberModel] = []
    var applicants: [GroupMemberModel] = []
    var leaderId: Int? = nil
    var capacity: Int = 0
    var currentMembers: Int = 0
    var isLoading: Bool = false
    var error: String? = nil
    var successMessage: String? = nil

    var isFull: Bool {
        currentMembers >= capacity && capacity > 0
    }
}

@Observable
@MainActor
final class MemberManageViewModel {
    private let groupRepository: GroupRepository
    private let groupService: GroupService
    var uiState = MemberManageUiState()

    init(groupRepository: GroupRepository, groupService: GroupService) {
        self.groupRepository = groupRepository
        self.groupService = groupService
    }

    func loadMembers(groupId: Int, groupKind: GroupKind) async {
        uiState.isLoading = true
        uiState.error = nil

        async let detailResult = loadGroupDetail(groupId: groupId, groupKind: groupKind)
        async let membersResult = loadGroupMembers(groupId: groupId, groupKind: groupKind)
        async let applicantsResult = loadApplications(groupId: groupId, groupKind: groupKind)

        let (detail, membersData, applicantsData) = await (detailResult, membersResult, applicantsResult)

        if let detail = detail {
            uiState.leaderId = detail.leaderId
            uiState.capacity = detail.capacity
            uiState.currentMembers = detail.currentMembers ?? 0
        }

        if let members = membersData {
            uiState.members = members
        }

        uiState.applicants = applicantsData

        uiState.isLoading = false
    }

    private func loadGroupDetail(groupId: Int, groupKind: GroupKind) async -> GroupDetailModel? {
        let result: Result<GroupDetailModel, Error>
        if groupKind == .study {
            result = await groupRepository.getStudyDetail(studyId: groupId)
        } else {
            result = await groupRepository.getTeamDetail(teamId: groupId)
        }
        return try? result.get()
    }

    private func loadGroupMembers(groupId: Int, groupKind: GroupKind) async -> [GroupMemberModel]? {
        let result: Result<[GroupMemberModel], Error>
        if groupKind == .study {
            result = await groupRepository.getStudyMembers(studyId: groupId)
        } else {
            result = await groupRepository.getTeamMembers(teamId: groupId)
        }

        switch result {
        case .success(let data):
            return data
        case .failure(let error):
            uiState.error = error.localizedDescription
            return nil
        }
    }

    private func loadApplications(groupId: Int, groupKind: GroupKind) async -> [GroupMemberModel] {
        do {
            let responses: [GroupApplicationDetailResponse]
            if groupKind == .study {
                responses = try await groupService.getStudyApplications(studyId: groupId)
            } else {
                responses = try await groupService.getTeamApplications(teamId: groupId)
            }
            return responses
                .filter { $0.status.uppercased() == "PENDING" }
                .map { $0.toApplicantModel() }
        } catch {
            // Applications may fail if not leader - silently return empty
            return []
        }
    }

    func acceptApplication(groupId: Int, groupKind: GroupKind, applicationId: Int) async {
        do {
            if groupKind == .study {
                try await groupService.acceptStudyApplication(applicationId: applicationId)
            } else {
                try await groupService.acceptTeamApplication(applicationId: applicationId)
            }
            uiState.successMessage = "멤버를 수락했습니다."
            await loadMembers(groupId: groupId, groupKind: groupKind)
        } catch {
            uiState.error = error.localizedDescription
        }
    }

    func rejectApplication(groupId: Int, groupKind: GroupKind, applicationId: Int) async {
        do {
            if groupKind == .study {
                try await groupService.rejectStudyApplication(applicationId: applicationId)
            } else {
                try await groupService.rejectTeamApplication(applicationId: applicationId)
            }
            uiState.successMessage = "지원을 거절했습니다."
            await loadMembers(groupId: groupId, groupKind: groupKind)
        } catch {
            uiState.error = error.localizedDescription
        }
    }

    func removeMember(groupId: Int, groupKind: GroupKind, memberId: Int) async {
        do {
            if groupKind == .study {
                try await groupService.removeStudyMember(studyId: groupId, memberId: memberId)
            } else {
                try await groupService.removeTeamMember(teamId: groupId, memberId: memberId)
            }
            uiState.successMessage = "멤버를 내보냈습니다."
            await loadMembers(groupId: groupId, groupKind: groupKind)
        } catch {
            uiState.error = error.localizedDescription
        }
    }

    func clearMessages() {
        uiState.error = nil
        uiState.successMessage = nil
    }
}

File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/mygroup/screen/MemberManageViewModel.kt
package com.ssafy.ssabree.features.mygroup.screen

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.core.repository.GroupRepository
import com.ssafy.ssabree.core.repository.MemberRepository
import com.ssafy.ssabree.core.repository.MyGroupRepository
import com.ssafy.ssabree.core.repository.model.GroupDetailModel
import com.ssafy.ssabree.core.repository.model.GroupMemberModel
import com.ssafy.ssabree.features.group.model.GroupKind
import com.ssafy.ssabree.features.mygroup.model.ApplicationUiModel
import com.ssafy.ssabree.features.mygroup.model.MemberUiModel
import com.ssafy.ssabree.features.mygroup.model.toUiModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope

data class MemberManageUiState(
    val applications: List<ApplicationUiModel> = emptyList(),
    val members: List<MemberUiModel> = emptyList(),
    val leaderId: Long? = null,
    val capacity: Int = 0,
    val memberCount: Int = 0,
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
    val successMessage: String? = null
)

class MemberManageViewModel(
    private val groupRepository: GroupRepository,
    private val repository: MyGroupRepository,
    private val memberRepository: MemberRepository,
    private val groupKind: GroupKind,
    private val groupId: Long
) : ViewModel() {
    private val _uiState = MutableStateFlow(MemberManageUiState())
    val uiState: StateFlow<MemberManageUiState> = _uiState.asStateFlow()

    init {
        load()
    }

    fun load() = viewModelScope.launch {
        _uiState.update { it.copy(isLoading = true, errorMessage = null) }
        val appsDeferred = async { repository.getApplications(groupKind, groupId) }
        val detailDeferred = async {
            if (groupKind == GroupKind.STUDY) {
                groupRepository.getStudyDetail(groupId)
            } else {
                groupRepository.getTeamDetail(groupId)
            }
        }
        val membersDeferred = async {
            if (groupKind == GroupKind.STUDY) {
                groupRepository.getStudyMembers(groupId)
            } else {
                groupRepository.getTeamMembers(groupId)
            }
        }

        val appsResult = appsDeferred.await()
        val detailResult = detailDeferred.await()
        val membersResult = membersDeferred.await()

        val error = listOf(
            appsResult.exceptionOrNull(),
            detailResult.exceptionOrNull(),
            membersResult.exceptionOrNull()
        ).firstOrNull { it != null }

        if (error != null) {
            _uiState.update { it.copy(isLoading = false, errorMessage = error.message) }
            return@launch
        }

        val mergedMembers = mergeLeader(
            membersResult.getOrNull().orEmpty(),
            detailResult.getOrNull()
        ).map { it.toUiModel() }
        val enrichedMembers = enrichMattermostIds(mergedMembers)
        val detail = detailResult.getOrNull()
        val memberCount = maxOf(detail?.currentMembers ?: 0, enrichedMembers.size)
        _uiState.update {
            it.copy(
                applications = appsResult.getOrNull()?.map { it.toUiModel() } ?: emptyList(),
                members = enrichedMembers,
                leaderId = detailResult.getOrNull()?.leaderId,
                capacity = detail?.capacity ?: 0,
                memberCount = memberCount,
                isLoading = false
            )
        }
    }

    fun accept(applicationId: Long) = viewModelScope.launch {
        repository.acceptApplication(groupKind, applicationId)
            .onSuccess { load() }
            .onFailure { e -> _uiState.update { it.copy(errorMessage = e.message) } }
    }

    fun reject(applicationId: Long) = viewModelScope.launch {
        repository.rejectApplication(groupKind, applicationId)
            .onSuccess { load() }
            .onFailure { e -> _uiState.update { it.copy(errorMessage = e.message) } }
    }

    fun kickMember(memberId: Long) = viewModelScope.launch {
        val result = when (groupKind) {
            GroupKind.PROJECT -> groupRepository.kickTeamMember(groupId, memberId)
            GroupKind.STUDY -> groupRepository.kickStudyMember(groupId, memberId)
        }
        result
            .onSuccess {
                _uiState.update { it.copy(successMessage = "멤버가 추방되었습니다.") }
                load()
            }
            .onFailure { e ->
                _uiState.update { it.copy(errorMessage = e.message) }
            }
    }

    fun clearErrorMessage() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    fun clearSuccessMessage() {
        _uiState.update { it.copy(successMessage = null) }
    }

    private fun mergeLeader(
        members: List<GroupMemberModel>,
        detail: GroupDetailModel?
    ): List<GroupMemberModel> {
        val leaderId = detail?.leaderId ?: return members
        if (members.any { it.id == leaderId }) return members
        val leader = GroupMemberModel(
            id = leaderId,
            email = detail.leaderEmail,
            name = detail.leaderName,
            mattermostId = detail.leaderMattermostId,
            profileImageUrl = detail.leaderProfileImageUrl
        )
        return members + leader
    }

    private suspend fun enrichMattermostIds(members: List<MemberUiModel>): List<MemberUiModel> = coroutineScope {
        members.map { member ->
            async {
                if (member.mattermostId.isNotBlank() && member.mattermostId != "-") {
                    member
                } else {
                    val mattermostId = memberRepository.getMember(member.id).getOrNull()?.mattermostId
                    if (!mattermostId.isNullOrBlank()) {
                        member.copy(mattermostId = mattermostId)
                    } else {
                        member
                    }
                }
            }
        }.awaitAll()
    }
}
